Use threading wherever necessary

fushinari opened the pull request:
Fixes #17 .

@rafaelmardojai There's a possible issue with this implementation. A minor race condition if the Translate button is clicked continuously. In testing, it has never happened but theoretically it can since I set the variables "ActiveThread" and "WaitingThread" from within the thread. I don't know how to fix it. Any suggestions?

rafaelmardojai commented:
Mmm, there isn't conditional code to check if the text to translate has changed at all? Because if the text is the same the Translate button should do nothing.

fushinari commented on src/dialect.py:
```diff
@@ -488,42 +500,61 @@ def Translation(self, button):
+                json.dump(self.Settings, outfile, indent=2)
+
+        self.ActiveThread = self.WaitingThread
+        self.WaitingThread = None
```
@rafaelmardojai This is the problem. I'm switching the active thread here. But the issue is that if the user clicks the Translate button while this is happening, L504 to L515 might get executed at conflicting times. I don't know what will be the end result but it will be bad.

rafaelmardojai replied:
Yeah, now i understand what you where saying. 

What about using the `queue` python module? I have never used it, but maybe will solve our issue.

rafaelmardojai commented:
What if we have a permanent ruining thread for the translation?

We have a list called `self.queue` where we store the strings to translate every.

When the user clicks the Translate button we add the text to the queue with something like:

```python
def add_to_queue(self):
    text =  # get self.LeftBuffer text
    # apply any conditional to the text
    old = self.queue.pop(0) # remove previous text from queue
    self.queue.append(text) # add text to queue
```
The permanent thread could look like:

```python
def Translation(self):
    while True:
        if self.queue:
            text = self.queue.pop(0)
            # Translate and do wherever with the text 
```

fushinari commented:
> What if we have a permanent ruining thread for the translation?
> 
> We have a list called `self.queue` where we store the strings to translate.
> 
> When the user clicks the Translate button we add the text to the queue with something like:
> 
> ```python
> def add_to_queue(self):
>     text =  # get self.LeftBuffer text
>     # apply any conditional to the text
>     old = self.queue.pop(0) # remove previous text from queue
>     self.queue.append(text) # add text to queue
> ```
> 
> The permanent thread could look like:
> 
> ```python
> def Translation(self):
>     while True:
>         if self.queue:
>             text = self.queue.pop(0)
>             # Translate and do wherever with the text 
> ```

Nice, this could work. Let me try.

fushinari commented:
@rafaelmardojai Can you test the current implementation? It's working well on my end. It doesn't do the cleaning of unnecessary requests anymore, but then again, maybe that was a bad idea anyway because it goes against how threads work.

fushinari commented:
Re-added the cleaning and just hope that the GIL means what I think it means, lol.

fushinari commented:
Is it just me or is gTTS not working anymore?

rafaelmardojai commented:
> Is it just me or is gTTS not working anymore?

Yeah, see: https://github.com/pndurette/gTTS/issues/226

rafaelmardojai commented:
> @rafaelmardojai Can you test the current implementation?

All looks good and smooth on my end.

We need to add some feedback to let the user know that something is happening when he click the Translate button. 
Making the `LeftText` insensitive? Adding a spinner somewhere?

fushinari commented:
> > @rafaelmardojai Can you test the current implementation?
> 
> All looks good and smooth on my end.
> 
> We need to add some feedback to let the user know that something is happening when he click the Translate button.
> Making the `LeftText` insensitive? Adding a spinner somewhere?

We could do the same thing as the audio button. Putting a spinner on the button I mean. But then, that could make the user assume that he/she can't click "Translate" again until it's done. I don't see any other logical position for the spinner to be in either...

rafaelmardojai commented:
Other issue is that the app takes a while to show up, probably because slow internet connection. Do we make a request to the API on startup?

fushinari commented:
> Other issue is that the app takes a while to show up, probably because slow internet connection. Do we make a request to the API on startup?

Yes, for language codes. The app will not even launch without internet. I want to change this, but a lot of the code depends on it.

rafaelmardojai commented:
> We could do the same thing as the audio button. Putting a spinner on the button I mean. But then, that could make the user assume that he/she can't click "Translate" again until it's done. I don't see any other logical position for the spinner to be in either...

Could we change the left text view with a spinner when translating? Or add some points (...) at the end of the left text and make the view insensitive as Google Translate.

rafaelmardojai commented:
> Yes, for language codes. The app will not even launch without internet. I want to change this, but a lot of the code depends on it.

Right, we should show a empty window with a spinner on startup and also have some error view.

fushinari commented:
> Right, we should show a empty window with a spinner on startup and also have some error view.

I did add an error dialog when no network is available. But a splash screen also seemed like a bad idea since no other GNOME app does it. Maybe use a stack with a spinner on one page and the actual content on another page and switch pages when loading is done?

fushinari commented:
> > We could do the same thing as the audio button. Putting a spinner on the button I mean. But then, that could make the user assume that he/she can't click "Translate" again until it's done. I don't see any other logical position for the spinner to be in either...
> 
> Could we change the left text view with a spinner when translating? Or add some points (...) at the end of the left text and make the view insensitive as Google Translate.

Then the ability to type while the current translation is in progress goes away. Or should we only keep that ability while in live mode?

rafaelmardojai commented:
> Then the ability to type while the current translation is in progress goes away. Or should we only keep that ability while in live mode?

LOL, Sorry, I wanted to say the RIGHT text. 😅

rafaelmardojai commented:
> I did add an error dialog when no network is available. But a splash screen also seemed like a bad idea since no other GNOME app does it. Maybe use a stack with a spinner on one page and the actual content on another page and switch pages when loading is done?
 
I was thinking something like my [Webfont Kit Generator](https://github.com/rafaelmardojai/WebfontKitGenerator) app progress view:
![3](https://user-images.githubusercontent.com/6210397/93163235-345d7700-f6dc-11ea-9a8f-8fa7aae2f7e0.png)

fushinari commented:
> LOL, Sorry, I wanted to say the RIGHT text. :sweat_smile:

Lol, it's fine. Adding a "..." to the end is easy, like Google Translate, but we might want to put off adding proper feedback for later. I mean, at this point, it's a given that we need to re-do the whole UI in Glade. Might be better to wait until then for the loading view too.

rafaelmardojai commented:
Yeah. I think this PR is ready to merge.

When merged I will work on an initial split of files.

fushinari commented:
> 
> I was thinking something like my [Webfont Kit Generator](https://github.com/rafaelmardojai/WebfontKitGenerator) app progress view:
> 
> So we should have three views:
> 
>   * Loading
>   * Network error
>   * Translator

Yep, this was almost what I had in mind too. But adding more UI elements in this code would be a waste of time. I can only do proper Handy-based development when GNOME 3.38 is out because my environment is a mess right now (libhandy and libhandy1 conflicts in Glade and such). That's my fault entirely but I haven't had the time to fix it. (The AUR `libhandy1` package is also mine and I haven't looked into all the issues it has... People are using it though... They don't know what I'm putting them through :sweat_smile: )

fushinari commented:
@gi-lom It's okay to merge now.

rafaelmardojai commented:
Just noticed that switching languages will freeze the UI, not too much noticeable, so we can fix it later.

fushinari commented:
> Just noticed that switching languages will freeze the UI, not too much noticeable, so we can fix it later.

Yes, there's a call to the translator there too. L325, Should I thread it?

rafaelmardojai commented:
> Should I thread it?

Ideally yes. It takes a while when the language is in Auto.

rafaelmardojai commented:
Also noticed `self.Translator` is defined twice, L53 & L86.

fushinari commented:
> Ideally yes. It takes a while when the language is in Auto.

I've fixed this and corrected the double definition. I'm noticing that the current history storage mechanism is unreliable. But that's for later.

rafaelmardojai commented:
> I've fixed this and corrected the double definition. I'm noticing that the current history storage mechanism is unreliable. But that's for later.

Yeah, I'm also suggesting using GSettings (#36)

rafaelmardojai commented:
@gi-lom merge?

gi-lom commented:
Sorry for the delay
